[MM-44796] Include participants in Recent, Saved and Pinned APIs responses.

vish9812 opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does, as well as QA test steps (if applicable and if not already added to the Jira ticket).
-->
Include participants in Recent, Saved and Pinned APIs responses.

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
Fixes https://mattermost.atlassian.net/browse/MM-44796

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
* Fixed the following 3 APIs to populate the "participants" field of the "post" response model.
  * posts/search
  * users/{user_id}/posts/flagged
  * channels/{channel_id}/pinned
```

ashishbhate reviewed (changes requested):
@vish9812 I did a quick first pass review. PR looks good in general. I've commented about a few things I'm not clear about. 

Also some tests to confirm that we're returning the extra data correctly would be good.

vish9812 commented on model/post.go:
```diff
@@ -117,6 +117,13 @@ type PostEphemeral struct {
 	Post   *Post  `json:"post"`
 }
 
+type PostWithExtra struct {
+	ThreadReplyCount   int64
+	IsFollowing        *bool
+	ThreadParticipants StringArray
+	Post
+}
```
>  The "extra" fields here are already in the Post struct

Yeah, it looks like only `ThreadParticipants` is needed in this model. The other two fields can be used from the `Post`. I can try removing the 2 fields but I guess there could be a reason why they were added in the first place. 😕 

> I'm not sure this is the best place for this struct

So far, I've only seen `model` package as a common place for all models.
Other alternative, I see is to import the `sqlstore` directly in the store and other layers packages, which I don't see can be a good thing.
So, I'm leaving it in `model` for now, LMK, if you've any other proposal.

> a better name

How about `ThreadedPost` as it's meant to be used in `PrepareThreadedResponse` method only?

ashishbhate replied:
> I can try removing the 2 fields but I guess there could be a reason why they were added in the first place.

I'm not sure why they are there in the first place either. You might need to look in git history or maybe other parts of the code. 


>So, I'm leaving it in model for now, LMK, if you've any other proposal.

SGTM

>How about ThreadedPost as it's meant to be used in PrepareThreadedResponse method only?

0/5 on the name, maybe something a more obvious like - `PostWithCRTMetadata`

vish9812 commented on store/sqlstore/channel_store.go:
```diff
@@ -807,17 +807,31 @@ func (s SqlChannelStore) InvalidateChannelByName(teamId, name string) {
 func (s SqlChannelStore) GetPinnedPosts(channelId string) (*model.PostList, error) {
-	pl := model.NewPostList()
+	builder := s.getQueryBuilder().Select(
+		"p.*",
+		"COALESCE(Threads.ReplyCount, 0) as ThreadReplyCount",
+		"COALESCE(Threads.LastReplyAt, 0) as LastReplyAt",
+		"COALESCE(Threads.Participants, '[]') as ThreadParticipants",
```
> we're adding some additional fields as well

Yeah, since there is already a common method `PrepareThreadedResponse` which updates all these fields, so it's better to utilize it and populate all of them.

> Don't we also need IsFollowing from threadMemberships?

It's not exactly needed, and as it depends on userId which isn't available in this method. It seems unnecessary to update this method chain upto API layer, to include the userId.

koox00 replied:
If the thread is known to the client it's OK to not have the `isFollowing` set since it will maintain the known value, but If it's not known yet would it be shown as un-followed?

vish9812 replied:
> but If it's not known yet would it be shown as un-followed?

Since it's an existing working code, I assume the current clients already have the necessary information and don't depend on this `isFollowing` flag. But yeah, in the future something new may come up that start using the `isFollowing` without knowing that it is never set.

ashishbhate replied:
>Since it's an existing working code, I assume the current clients already have the necessary information and don't depend on this isFollowing flag

That's not necessarily a good assumption about our code 😛 

I checked in the webapp, pinned posts that I'm already following show up as unfollowed to me (it shows the "Follow" button, which means the app thinks I'm not following the thread)

ashishbhate reviewed (commented):
Commented inline. It would be good to have tests to confirm that the recent, saved and pinned post API responses now have the correct additional data

ashishbhate reviewed (commented):
- We should add tests for flagged posts as well. The changes in the query are complex enough that we should throughly test the changes. 

- The JIRA ticket and the summary for this PR say that we're adding this information for searches as well (recent mentions), but I don't see the changes for that here. I'm not sure if I'm missing something.

vish9812 commented:
* I already added tests for Flagged Posts in `store/storetest/post_store.go#2658`. But there was test missing for search, which I've added now in `store/searchtest/post_layer.go#1851`.
* As for search and recent, both are same thing from code perspective. Recent just calls search.

ashishbhate reviewed (commented):
Awesome work! 

Have you been able to load test the changes? In this PR there are queries that do a join with the posts table. In the past this has caused performance issues. 

I suggest:

1. Confirming the loadtest tool already covers the recent mentions, saved and pinned API calls. If not, please add them, this should be straightforward to do. 
2. Run the loadtest on this branch and on master and then compare the results.

Willyfrog commented:
@vish9812 did you manage to load test this? is it blocked by something else?

vish9812 commented:
I'm waiting for @koox00 to finish his load-tests. Then I can setup a call with him, to help me out.

isacikgoz commented:
@vish9812 is this PR active? If not can you please close it? Thanks!

vish9812 commented:
Thanks for pinging on this one @isacikgoz. I'll validate if it's still an issue and will update it accordingly this week.

larkox commented:
@agarciamontoro Heads up that this may need load testing (some changes to APIs behaviour at store level).

agnivade commented:
Hey @larkox, Alejandro is currently super focussed with the ceiling tests. If this needs him to run the tests, then it probably needs to wait for a while. Otherwise, Vishal has also some experience running a load test I believe.

larkox commented:
@agnivade not sure if there is a need for this pr to go through the load testing, and not sure what is the process here. I tagged Alejandro since he is my go to person for load test :sweat_smile: I mainly wanted to flag this PR for you guys, because it deals with changes on the APIs, and may or may not have an impact on performance (therefore the load test).

vish9812 commented:
It should be load-tested as it adds more data to the apis responses.
And I already plan to load-test it. :smiley:

agarciamontoro commented:
Thanks, @vish9812!
